### DISCUSSION TITLE
github will require 2FA

### DISCUSSION THREAD

**Discussion by stsp** [dosemu2/dosemu2#2074] [Off-topic]
Today I've got the e-mail that in 45
days I'll need to enable the two-factor authentication,
or my access to an account will be
restricted.
Since I am not going to give github
my phone number, I wonder what to
do now. Does anyone know the details?
Will there be some other way around,
or should I start moving dosemu2 to
another hosting right now?

**Comment by bartoldeman** [comment]
You can also use a authenticator app on your phone or desktop, eg Authy, no need for phone numbers or sms.

**Comment by jschwartzenberg** [comment]
I use this Firefox add-on: https://addons.mozilla.org/en-US/firefox/addon/foxauth/

**Comment by stuaxo** [comment]
I use an autheniticator app on my phone, with some backup codes printed (somewhere?).   I do also do the phone number thing.

**Comment by stsp** [comment]
Thanks guys!
I don't think I'll use phone-based
solutions, but desktop-based look
good too.

**Comment by jharrison03** [comment]
You don't need to use a phone at all.

You can use a FIDO U2F token (a "Yubikey"-type device) which never needs to see a phone.

You just need the hardware token and a recent Firefox or Chromium on Linux to register the keys.

**Reply by stsp** [reply]
It seems, according to the above,
the HW key is not even needed.
I think they can authenticate you
with a private key stored on notebook,
which should work for me.

**Reply by jschwartzenberg** [reply]
> I think they can authenticate you
> with a private key stored on notebook,

Yes you get a key that you should not expose, similarly to a password. You can copy this secret key around as much as you like, just like you could with a password. The system is called TOTP. The same key is kept on the other side. You then use a TOTP calculator which takes the secret key plus the system time and calculates a 6 digit code. Entering this code when logging in proves you have a copy of that secret key as the other side can do the same comparison.
The main thing to keep in mind is that the time is accurate/synchronized on any device which calculates the 6 digits.

Typically the key is provided in BASE32, but some tools can also use hexademical syntax. You can feed the command line tool `oathtool` with the key and the right algorithm to get these 6 digits on the command line. You can find it in the repositories of all major GNU/Linux distros. There are also various other TOTP calculators in the form of browser add-ons and so on which can also help to store the actual secrets for you. Typically programs that store these secrets and combine them with a TOTP calculator so you can see just the current 6 digit code are called "authenticators".

**Reply by stsp** [reply]
> The main thing to keep in mind is that the time is accurate/synchronized on any device which calculates
> the 6 digits.

I expected it to use the push notification
with the random value. Strange that they
use system time for that.

**Reply by jschwartzenberg** [reply]
The idea is that you only need a device that accurately tracks time, possibly through a time signal. This allows one to keep the secret exclusively on a device without an internet connection and let that device only (conveniently) expose those 6 digit codes.

**Reply by stsp** [reply]
I wonder if that's true, because in
that case not only you need to keep
time in sync between both ends, but
also generate the digest at the very
precise same moment, or it will mismatch.
I don't understand how can that be
done w/o internet connection.
And also if your "device" is a notebook
with firefox plugin, and you log to github
on exactly that device, then you need
internet at least to enter the code, so
pretending there is no internet connection,
is in that case at least strange.

**Reply by jschwartzenberg** [reply]
> very precise same moment

Yes, but keep in mind that "moment" in this case is 30 seconds. Usually systems also accept the code of the previous moment for about 15-30 seconds I guess and maybe also the next code already for about 15-30 seconds. Only every 30 seconds the TOTP calculation will result in a new code.

> And also if your "device" is a notebook
with firefox plugin, and you log to github
on exactly that device, then you need
internet at least to enter the code, so
pretending there is no internet connection,
is in that case at least strange.

Entering the code will require a connection. But the device that stores the secret and generates the code can be a different device and that different device would not need an internet connection. The idea is to be able to always keep the secret on an offline device.

**Comment by andrewbird** [comment]
Mmm, slightly pissed off here as github thinks I now need 2 factor authentication. Surely that's my decision, it's not like I can cost them money if somebody steals my password and impersonates me. Not sure I can be bothered with it all.

**Reply by stsp** [reply]
Hi Andrew!
Same attitude here, but the end of
the day I just installed the Totp firefox
plugin and the QR Code scanner plugin,
and spent 5 minutes figuring out what
and where is to copy/paste. :)
So its really not a big deal.
Just spend 5 minutes and forget.

**Reply by andrewbird** [reply]
I gave in and used a prepay phone for SMS.

**Reply by jharrison03** [reply]
Don't use SMS... Use an authenticator app with TOTP, at least.  If your pre-paid phone is Android, consider Authenticator Pro.  In the US at least, you don't even need a SIM for the phone to get time-sync to generate the OTP.

Using HW keys though, you don't need any phone.  The keys are registered directly with the site, like GH, and you can use them with any other sites that support FIDO U2F.  The support for keys are built-into the browser.
